src/workerd/server/server.c++

```diff
@@ -2494,6 +2497,94 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
     isolateRegistrar->registerIsolate(name, isolate.get());
   }
 
+  if (conf.hasModuleFallback()) {
+    KJ_REQUIRE(experimental,
+               "The module fallback service is an experimental feature. "
+               "You must run workerd with `--experimental` to use this feature.");
+    // If the config has the moduleFallback option, then we are going to set up the ability
+    // to load certain modules from a fallback service. This is generally intended for local
+    // dev/testing purposes only.
+    auto& apiIsolate = isolate->getApiIsolate();
+    apiIsolate.setModuleFallbackCallback(
+        [address=kj::str(conf.getModuleFallback()), featureFlags=apiIsolate.getFeatureFlags()]
+        (jsg::Lock& js,
+         kj::StringPtr specifier,
+         jsg::CompilationObserver& observer,
+         jsg::ModuleRegistry::ResolveMethod method) mutable
+            -> kj::Maybe<jsg::ModuleRegistry::ModuleInfo> {
+      kj::Maybe<kj::String> jsonPayload;
+      {
+        // Module loading in workerd is expected to be synchronous but we need to perform
+        // an async HTTP request to the fallback service. To accomplish that we wrap the
+        // actual request in a kj::Thread, perform the GET, and drop the thread immediately
+        // so that the destructor joins the current thread (blocking it). The thread will
+        // either set the jsonPayload variable or not.
+        kj::Thread loaderThread(
+          [specifier,address=address.asPtr(),&jsonPayload, method]() mutable {
+          try {
+            const auto toStr = [](jsg::ModuleRegistry::ResolveMethod method) {
+              switch (method) {
+                case jsg::ModuleRegistry::ResolveMethod::IMPORT: return "import"_kjc;
+                case jsg::ModuleRegistry::ResolveMethod::REQUIRE: return "require"_kjc;
+              }
+              KJ_UNREACHABLE;
+            };
+
+            kj::AsyncIoContext io = kj::setupAsyncIo();
+
+            kj::HttpHeaderTable::Builder builder;
+            kj::HttpHeaderId kMethod = builder.add("resolve-method");
+            auto headerTable = builder.build();
+
+            auto addr = kj::heap<PromisedNetworkAddress>(
+                io.provider->getNetwork().parseAddress(address, 80));
+
+            auto client = kj::newHttpClient(
+                io.provider->getTimer(),
+                *headerTable,
+                *addr, {});
+
+            kj::HttpHeaders headers(*headerTable);
+            headers.set(kMethod, toStr(method));
+            headers.set(kj::HttpHeaderId::HOST, "localhost"_kj);
+
+            auto request = client->request(
+                kj::HttpMethod::GET,
+                kj::str("file://", specifier),
```

Comment:
An HTTP request path should start with `/`. A protocol prefix like `file://` is not expected here.

This code as-is will produce an HTTP request line like:

```
GET file://blah HTTP/1.1
```

Reply:
Yes, that's intentional here. The target identifier portion of the HTTP request line is a `URI Reference` which can be a fully-qualified absolute URL.

Reply:
FWIW the HTTP spec says you "MUST" use origin form (path, not URL) when talking to an origin server, and "MUST" use absolute form when talking to a proxy.

https://www.ietf.org/archive/id/draft-ietf-httpbis-messaging-11.html#name-origin-form

Admittedly I generally feel that if the client and server want to agree on something else, they are free to do so, regardless of what MUSTs appear in the spec. But it does seem clear that the HTTP spec does not intend for full URLs to be used in this context.

Reply:
The paragraphs you quote are interesting and I didn't realize origin servers were expected to support this. However, these pagagraphs seem to very clearly imply that the request-target should be an HTTP URL and that the hostname within that URL should be the hostname of the origin server. Sending `file://` seems clearly not in the spirit of what they're saying here.

I think the only way I could be convinced that this is an appropriate way to use the HTTP protocol is if you really go all in on arguing that this is a proxy. But even then, this is a very... unusual design. I would guess that a lot of HTTP server frameworks won't even support these kinds of requests. I suppose since it's a private protocol, the only HTTP server that matters is whatever miniflare uses. But it feels like asking for trouble.